### PR TITLE
feat: implement type safety

### PR DESCRIPTION
**Pull request by TomerAberbach**
- Implement type-level checking of graph constraints
- The above requires an API change, to separate the keys representing a task from the task function
- Delete debug runtime validation entrypoint since the type-level validation takes care of it
- Reduce bundle size
- Increase test coverage
- Deprecate `grfnviz` since it's a pain to maintain and no one seems to be using it
- Require node 18

### CODE REVIEW

**Comment by TomerAberbach** [comment]
@maraisr since you're one of the people I know uses this, do you have any feedback? Does this look good to you at a high-level?

**Comment by maraisr** [comment]
This is such an wonderful pull request! 🎉 Thank you so much!!

Was messing around with it, and stumbled upon this nugget. Being able to detect a single output vertex is such a major win for stability and assurance.

image

I'd say ship it! :shipit: 🚀 

My only critique (though pales in comparison to the wins) is that I was unable to get autocomplete for dependencies, but its no biggy.

<details><summary>Example</summary>

```ts
let get = grfn({
	get_post,
	get_comments: [get_comments, ['get_post']],
	get_author: [get_author, ['get_post']],
	format: [
		(comments: string, post: string, author: string) => {
			return {
				comments,
				post,
				author
			};
		},
		['get_comments', 'get_post', 'get_au ▌']
//                                                   ^ expected intellisense to provide me with "get_author"
	]
});

let result = await get('my-post');
console.log(`result: %o`, result);
```

</details>

File: readme.md [hunk 1]
```diff
@@ -13,24 +13,26 @@
...
     <img src="https://github.com/TomerAberbach/grfn/workflows/CI/badge.svg" alt="CI" />
   </a>
-  <a href="https://bundlephobia.com/result?p=grfn">
-    <img src="https://badgen.net/bundlephobia/minzip/grfn" alt="minzip size" />
+  <a href="https://unpkg.com/grfn/dist/index.min.js">
+    <img src="https://deno.bundlejs.com/?q=grfn&badge" alt="gzip size" />
+  </a>
+  <a href="https://unpkg.com/grfn/dist/index.min.js">
+    <img src="https://deno.bundlejs.com/?q=grfn&config={%22compression%22:{%22type%22:%22brotli%22}}&badge" alt="brotli size" />
   </a>
```
Comment (maraisr) [inline readme.md:21]:
```suggestion
  </a>
  <a href="https://licenses.dev/npm/grfn">
    <img src="https://licenses.dev/b/npm/grfn" alt="licenses" />
  </a>
```

Nice to show off your licenses 😎

**Comment by TomerAberbach** [comment]
> This is such an wonderful pull request! 🎉 Thank you so much!!
> 
> Was messing around with it, and stumbled upon this nugget. Being able to detect a single output vertex is such a major win for stability and assurance.
> 
> image
> 
> I'd say ship it! :shipit: 🚀
> 
> My only critique (though pales in comparison to the wins) is that I was unable to get autocomplete for dependencies, but its no biggy.
> 
> Example

Glad it's working well for you! Agreed that autocomplete for dependencies would be nice. Might very well be possible to support...

**Pull request merged**